### PR TITLE
sdr-pipeline: Stream, Block, Chain, Splitter

### LINKED ISSUE #14: sdr-pipeline: Stream<T> double-buffer swap channel
## Description

Port the core streaming primitive -- the double-buffer swap channel that all inter-block data flows through.

## C++ Reference

- `core/src/dsp/stream.h` -- `dsp::stream<T>` template class
- Two 1M-sample buffers (`writeBuf`, `readBuf`) swapped atomically under mutex + condvar
- `swap(size)` -- writer swaps buffers, blocks if previous data not consumed
- `read()` -- reader waits for data, returns sample count
- `flush()` -- reader signals it has consumed the data
- `stopWriter()` / `stopReader()` -- graceful shutdown signals

## Deliverables

- `sdr-pipeline/src/stream.rs`: `Stream<T: Copy + Send>`, `StreamWriter<T>`, `StreamReader<T>`
- Pre-allocated 1M sample buffers (configurable via `STREAM_BUFFER_SIZE`)
- Mutex + Condvar synchronization matching C++ semantics exactly
- `swap(count)`, `read() -> Result<usize>`, `flush()`, `stop_writer()`, `stop_reader()`
- Zero allocation in hot path
- Unit tests: single-producer/single-consumer throughput, stop semantics, backpressure

### LINKED ISSUE #15: sdr-pipeline: Block trait and processor wrappers
## Description

Implement the threaded processing block abstraction.

## C++ Reference

- `core/src/dsp/block.h` -- Base block with `start()`, `stop()`, `workerLoop()` thread
- `core/src/dsp/processor.h` -- Processor with input/output streams
- `core/src/dsp/hier_block.h` -- Hierarchical composite block

## Deliverables

- `sdr-pipeline/src/block.rs`: `Block` trait (start/stop), `Processor<I, O>` trait
- `ProcessorBlock<F>` wrapper that spawns a worker thread calling a processing function in a loop
- `SourceBlock` / `SinkBlock` wrappers for source and sink modules
- Thread lifecycle management (join on stop, panic propagation)
- Unit tests with mock processors verifying start/stop/data flow

### LINKED ISSUE #16: sdr-pipeline: Chain and Splitter
## Description

Implement the processor chain and stream fan-out splitter.

## C++ Reference

- `core/src/dsp/chain.h` -- Ordered list of processors with enable/disable/rewire
- `core/src/dsp/routing/splitter.h` -- Fan-out one stream to multiple consumers

## Deliverables

- `sdr-pipeline/src/chain.rs`: `Chain<T>` with add/remove/enable/disable processors dynamically
- When a processor is disabled, chain rewires around it (next processor reads from previous)
- `sdr-pipeline/src/splitter.rs`: `Splitter<T>` that copies input to N output streams
- Unit tests: chain with 3 processors, disable middle one, verify data flows correctly

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

Pipeline infrastructure foundations — the streaming and processing backbone.

- `Stream<T>`: double-buffer swap channel with Mutex+Condvar backpressure, stop semantics
- `BlockRunner`: worker thread wrapper with RAII stop-on-drop, cooperative shutdown via AtomicBool
- `Chain<T>`: ordered named processors with enable/disable, ping-pong internal buffers
- `Splitter<T>`: fan-out copies input to N output buffers with dynamic add/remove

## Test plan

- [x] 24 unit tests passing (`cargo test -p sdr-pipeline`)
- [x] `cargo clippy --all-targets --workspace -- -D warnings` clean
- [x] `cargo fmt --all -- --check` clean
- [x] Stream: swap/read/flush cycles, stop writer/reader
- [x] Block: runner starts, stops cooperatively, drop-stops, early stop
- [x] Chain: single/chained processors, disable bypass, enable toggle, all-disabled passthrough
- [x] Splitter: single/multiple outputs, independent copies, add/remove

Closes #14
Closes #15
Closes #16

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Modular DSP pipeline: controllable processing units with start/stop, automatic runner shutdown on drop.
  * Ordered processing chain: named, enable/disable steps, passthrough when all disabled, runtime step management and safety checks.
  * Thread-safe double-buffered stream: blocking swap/read, flush, capacity, and stop/unblock controls.
  * Fan-out splitter: duplicate inputs across dynamic outputs.
  * Public crate modules exported.

* **Tests**
  * Unit tests covering core behaviors and edge cases for all components.

### CODE REVIEW

File: crates/sdr-pipeline/src/chain.rs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        if enabled_count == 0 {
+            // No processors enabled — passthrough
+            output[..input.len()].copy_from_slice(input);
+            return Ok(input.len());
+        }
+
+        // Process through enabled steps using ping-pong buffers (a ↔ b)
+        self.buf_a.resize(input.len(), T::default());
+        self.buf_b.resize(input.len(), T::default());
+        self.buf_a[..input.len()].copy_from_slice(input);
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/chain.rs:150]: Fixed — internal buffers now sized to output.len() to support interpolating processors that produce more samples than input.

File: crates/sdr-pipeline/src/splitter.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    pub fn remove_output(&mut self, index: usize) -> Result<(), DspError> {
+        if index >= self.outputs.len() {
+            return Err(DspError::InvalidParameter(format!(
+                "output index {index} out of bounds (have {})",
+                self.outputs.len()
+            )));
+        }
+        self.outputs.remove(index);
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/splitter.rs:48]: Documented — added note that remove_output shifts subsequent indices and callers should not cache indices across removals.

File: crates/sdr-pipeline/src/stream.rs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        }
+
+        {
+            let mut state = self.state.lock().unwrap_or_else(PoisonError::into_inner);
+            state.data_ready = true;
+        }
+        self.ready_cv.notify_all();
+
+        true
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs:95]: Acknowledged — the &mut self on write_buf/swap is intentional for the current design where a single BlockRunner owns the Stream. True SPSC split() is a valid future enhancement but not needed yet since blocks own their streams. The thread-safe claim refers to the Mutex/Condvar synchronization for the read/flush side.

File: crates/sdr-pipeline/src/stream.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        {
+            let mut state = self.state.lock().unwrap_or_else(PoisonError::into_inner);
+            state.data_ready = false;
+        }
+        {
+            let mut state = self.state.lock().unwrap_or_else(PoisonError::into_inner);
+            state.can_swap = true;
+        }
+        self.swap_cv.notify_all();
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs:123]: Fixed — consolidated flush() into a single lock acquisition for both data_ready and can_swap updates.

File: crates/sdr-pipeline/src/stream.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs:95]: yeah - we will leave as is for now

File: crates/sdr-pipeline/src/stream.rs [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        }
+
+        {
+            let mut state = self.state.lock().unwrap_or_else(PoisonError::into_inner);
+            state.data_ready = true;
+        }
+        self.ready_cv.notify_all();
+
+        true
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs:95]: Fixed — swap() now uses single lock acquisition for all state updates, matching the flush() consolidation.

File: crates/sdr-pipeline/src/chain.rs [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+            if !step.enabled {
+                continue;
+            }
+            if src_is_a {
+                current_count = (step.processor)(&self.buf_a[..current_count], &mut self.buf_b)?;
+            } else {
+                current_count = (step.processor)(&self.buf_b[..current_count], &mut self.buf_a)?;
+            }
+            src_is_a = !src_is_a;
+        }
```
Comment (jasonherald) [inline crates/sdr-pipeline/src/chain.rs:171]: Fixed — added defensive check: if processor returns count > buf_size, returns BufferTooSmall error instead of panicking.

File: crates/sdr-pipeline/src/splitter.rs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+    /// Process input by copying to all output buffers.
+    ///
+    /// Each output buffer is resized to hold the input data.
+    pub fn process(&mut self, input: &[T]) {
+        for output in &mut self.outputs {
+            output.resize(input.len(), T::default());
+            output.copy_from_slice(input);
+        }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/splitter.rs:62]: Acknowledged — resize is a no-op after warmup with consistent block sizes. The current pattern is simpler and avoids the clear+extend overhead of resetting length to 0 on every call.

File: crates/sdr-pipeline/src/stream.rs [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+                .unwrap_or_else(PoisonError::into_inner);
+
+            if state.writer_stop {
+                return false;
+            }
+
+            state.data_size = size;
+            std::mem::swap(&mut self.write_buf, &mut self.read_buf);
+            state.can_swap = false;
+            state.data_ready = true;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs:90]: Fixed — added assert that swap size <= capacity. Panics early with descriptive message instead of silently storing an invalid data_size.

File: crates/sdr-pipeline/src/stream.rs [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+        let state = self
+            .ready_cv
+            .wait_while(state, |s| !s.data_ready && !s.reader_stop)
+            .unwrap_or_else(PoisonError::into_inner);
+
+        if state.reader_stop {
+            return -1;
+        }
+
+        state.data_size as i32
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs:112]: Acknowledged — the i32 return matches the SDR++ API exactly. Changing to Result<usize, StreamError> is cleaner Rust but would cascade changes through all block implementations. Will consider for a future refactor pass.

File: crates/sdr-pipeline/src/chain.rs [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+    /// Number of steps in the chain.
+    pub fn len(&self) -> usize {
+        self.steps.len()
+    }
+
+    /// Whether the chain has no steps.
+    pub fn is_empty(&self) -> bool {
+        self.steps.is_empty()
+    }
+
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/chain.rs]: Fixed — added remove() method for dynamic chain rewiring.

File: crates/sdr-pipeline/src/stream.rs [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    /// Swap the write buffer with the read buffer, publishing `size` samples.
+    ///
+    /// Blocks until the consumer has flushed the previous data.
+    /// Returns `false` if the writer was stopped (graceful shutdown).
+    pub fn swap(&mut self, size: usize) -> bool {
+        assert!(
+            size <= self.write_buf.len(),
+            "swap size ({size}) exceeds buffer capacity ({})",
+            self.write_buf.len()
+        );
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs]: Fixed — replaced assert! with return false for invalid sizes. Consistent with writer-stop behavior, no panic in library code.

File: crates/sdr-pipeline/src/chain.rs [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+    #[test]
+    fn test_buffer_too_small() {
+        let mut chain: Chain<f32> = Chain::new();
+        let input = [1.0, 2.0, 3.0];
+        let mut output = [0.0_f32; 2];
+        assert!(chain.process(&input, &mut output).is_err());
+    }
+}
```
Comment (jasonherald) [inline crates/sdr-pipeline/src/chain.rs:416]: Fixed — added test_remove_processor and test_remove_not_found tests.

File: crates/sdr-pipeline/src/stream.rs [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+    /// Get a reference to the read buffer.
+    pub fn read_buf(&self) -> &[T] {
+        &self.read_buf
+    }
+
+    /// Swap the write buffer with the read buffer, publishing `size` samples.
+    ///
+    /// Blocks until the consumer has flushed the previous data.
+    /// Returns `false` if the writer was stopped (graceful shutdown).
+    pub fn swap(&mut self, size: usize) -> bool {
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs:71]: Fixed — doc now documents both false paths: writer stopped or invalid size (zero/exceeds capacity).

File: crates/sdr-pipeline/src/chain.rs [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+    pub fn add<F>(&mut self, name: &str, processor: F)
+    where
+        F: FnMut(&[T], &mut [T]) -> Result<usize, DspError> + Send + 'static,
+    {
+        self.steps.push(ChainStep {
+            processor: Box::new(processor),
+            enabled: true,
+            name: name.to_string(),
+        });
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/chain.rs]: Fixed — Chain::add now returns Result and rejects duplicate names.

File: crates/sdr-pipeline/src/stream.rs [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    #[test]
+    fn test_stop_reader() {
+        let s: Stream<f32> = Stream::with_capacity(256);
+        s.stop_reader();
+        assert_eq!(s.read(), -1, "read should return -1 when reader stopped");
+        s.clear_read_stop();
+    }
+}
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs:254]: Fixed — added test_swap_zero_size_returns_false and test_swap_exceeds_capacity_returns_false.

File: crates/sdr-pipeline/src/chain.rs [hunk 16]
```diff
@@ -0,0 +1,384 @@
...
+
+    #[test]
+    fn test_buffer_too_small() {
+        let mut chain: Chain<f32> = Chain::new();
+        let input = [1.0, 2.0, 3.0];
+        let mut output = [0.0_f32; 2];
+        assert!(chain.process(&input, &mut output).is_err());
+    }
+}
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/chain.rs:416]: Fixed — added test_add_duplicate_name verifying duplicate rejection.

File: crates/sdr-pipeline/src/stream.rs [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+        s.clear_write_stop();
+    }
+
+    #[test]
+    fn test_stop_reader() {
+        let s: Stream<f32> = Stream::with_capacity(256);
+        s.stop_reader();
+        assert_eq!(s.read(), -1, "read should return -1 when reader stopped");
+        s.clear_read_stop();
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/stream.rs:253]: Fixed — stop/clear tests now verify that swap/read resume successfully after clearing the stop flag.

File: crates/sdr-pipeline/src/chain.rs [hunk 18]
```diff
@@ -0,0 +1,403 @@
...
+                current_count = (step.processor)(&self.buf_b[..current_count], &mut self.buf_a)?;
+            }
+            if current_count > buf_size {
+                return Err(DspError::BufferTooSmall {
+                    need: current_count,
+                    got: buf_size,
+                });
+            }
+            src_is_a = !src_is_a;
+        }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/chain.rs:171]: Fixed — added test_processor_overflow_detected exercising the defensive count check.

File: crates/sdr-pipeline/src/chain.rs [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+        );
+    }
+
+    #[test]
+    fn test_processor_overflow_detected() {
+        let mut chain: Chain<f32> = Chain::new();
+        // Processor that claims to have written more than buffer size
+        chain
+            .add("overflow", |_: &[f32], _: &mut [f32]| Ok(999_999))
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/chain.rs:403]: Fixed — extracted to OVERSIZED_COUNT named constant.
